File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        hmap = {')':'(', ']':'[', '}':'{'}
        if len(s) % 2 != 0:
            return False
        for i in s:
            if i not in hmap:
                stack.append(i)
            else:
                if not stack:
                    return False
                if hmap[i] == stack[-1]:
                    stack.pop()
                else:
                    return False
        return len(stack)==0

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        // valid parenthsis in my way 
        Stack <Character> st = new Stack<>();

        for(char ch : s.toCharArray())
        {
            if(ch=='(' || ch=='{' || ch=='['){
                st.push(ch);

            }
            // on removal 
            else{
                if(st.isEmpty())return false;
                else if(
                    ch==')' && st.peek()=='(' ||
                    ch=='}' && st.peek()=='{' ||
                    ch==']' && st.peek()=='[' 
                ){
                    st.pop();

                }else{
                    return false;
                }

            }
        }
        return st.isEmpty();
        
    }
    
}